### ISSUE TITLE
[selectors] Drop `:target-within`, deprecate `:focus-within`

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#8357]
`:target-within` and `:focus-within` were added to the language during an era when we thought we could never do `:has()`. They are special cases of `:has()`, namely to `:has(:target)` and `:has(:focus)`. Now that we have `:has()`, there's no reason to have them in the language. `:target-within` has no implementation, so we could just drop it. `:focus-within` is widely used (~[8% of websites](https://docs.google.com/spreadsheets/d/1OU8ahxC5oYU8VRryQs9BzHToaXcOntVlh6KUHjm15G4/edit#gid=370286500)) but it may make sense to deprecate so we could drop it one day, if usage drops.

**Comment by Loirooriol** [comment]
Just pointing out that

 - `:focus-within` includes the element itself, so it may be convenient to avoid `:is(:focus, :has(:focus))`
 - `:focus-within` includes descendants in the flat tree, unlike `:has(:focus)`
 - It's likely that `:focus-within` has better performance

**Comment by LeaVerou** [comment]
> `:focus-within` includes descendants in the flat tree, unlike `:has(:focus)`

Oh that's a very good point.

**Comment by fantasai** [comment]
@LeaVerou do you still want to pursue this, or is @Loirooriol's response good enough to retract it?

**Comment by LeaVerou** [comment]
> @LeaVerou do you still want to pursue this, or is @Loirooriol's response good enough to retract it?

There's still the part about `:target-within`.

**Comment by dbaron** [comment]
I think the other value of these selectors is that they are much faster than `:has()` (though they require the browser to do a little bit of extra up-front work to get that benefit).  I'm still quite concerned that `:has()` is going to create selector performance problems that will require things like lints to forbid `:has()` and developers to know to avoid it.

While in theory we could detect and optimize some very specific cases of `:has()` such as `:has(:target)`, that seems like it would create more confusion around what the performance characteristics of CSS selectors are.  There's value in creating a platform that is understandable to the developers that use it -- and this does apply to its performance characteristics.

**Comment by LeaVerou** [comment]
> While in theory we could detect and optimize some very specific cases of `:has()` such as `:has(:target)`, that seems like it would create more confusion around what the performance characteristics of CSS selectors are. There's value in creating a platform that is understandable to the developers that use it -- and this does apply to its performance characteristics.

There is definitely value in creating a platform that is understandable to the developers that use it; however my understanding was that one of the core design principles of CSS these days is that authors should not need to worry about performance at all; if something cannot be implemented performantly, it's not implemented at all. Is this not the case with `:has()`? I thought  the whole reason `:has()` became feasible was that it was shown it could be implemented with reasonable performance. Is this not the case?

**Comment by dbaron** [comment]
I believe the performance of `:has()` is significant in some cases, although that belief is based on conversations with implementers rather than measurement.  In particular, I see occasional suggestions for things like implementing a dark theme checkbox using `:root:has(#dark-theme-control:checked) ...` selectors, and my understanding is that a significant number of `:root:has(...)` selectors on a large page would add up to a measurable performance problem.

**Comment by LeaVerou** [comment]
I suspect once such usage of `:has()` becomes more widespread, there will be further optimizations for common patterns. We should not be designing CSS syntax around transient performance issues. Especially since such messaging takes a very long time to go away — there are devs even today who hesitate to use descendant selectors because of 2010 era guidance around selector performance.

**Comment by bkardell** [comment]
Just want to add more than a thumbs up here and re-state that I don't think we should deprecate `:focus-within` for all of the same reasons @Loirooriol mentioned above.  It means something different than any `:has()` and also,  (but not the main reason) because it is very specifically about focus it can also be very efficient at that.

**Comment by Mouvedia** [comment]
> RESOLVED: Keep :focus-within, replace :target-within with a note that we hope :has(:target) is sufficient for use cases

So `:focus-within` _is not_ deprecated?

**Comment by SebastianZ** [comment]
Correct, for the reasons outlined by @Loirooriol in https://github.com/w3c/csswg-drafts/issues/8357#issuecomment-1404939311.

Sebastian
